tests/test_multipart.py

```diff
@@ -848,6 +848,11 @@ class CustomReader(aiohttp.MultipartReader):
          b'\r\n') == bytes(buf))
 
 
+async def test_writer_write_no_parts(buf, stream, writer) -> None:
+    await writer.write(stream)
+    assert b'--:--\r\n' == bytes(buf)
```

Comment:
I wonder if there should be just closed boundary or pair of open-close ones. Need to check how others handles this case. Python's MIMEMultipart does the former for instance.

Reply:
a pair of open-close would mean a single empty part
just a close boundary means no parts

they are two different things, so i think this is the right approach

Reply:
Yes, they are different, but that's how stdlib parser acts on empty list and that's what RFC defines about multipart:
```
multipart-body := preamble 1*encapsulation 
               close-delimiter epilogue 

encapsulation := delimiter CRLF body-part 

delimiter := CRLF "--" boundary   ; taken from  Content-Type 
field. 
                               ;   when   content-type    is 
multipart 
                             ; There must be no space 
                             ; between "--" and boundary. 

close-delimiter := delimiter "--" ; Again, no  space  before 
"--" 

preamble :=  *text                  ;  to  be  ignored  upon 
receipt. 

epilogue :=  *text                  ;  to  be  ignored  upon 
receipt. 

body-part = <"message" as defined in RFC 822, 
         with all header fields optional, and with the 
         specified delimiter not occurring anywhere in 
         the message body, either on a line by itself 
         or as a substring anywhere.  Note that the 
         semantics of a part differ from the semantics 
         of a message, as described in the text.> 
```
Encapsulation with open boundary is required, otherwise it becomes malformed data.

Reply:
Currently, we didn't have any issues of bad multipart reading because no data, but I guess we could meet such one. That's why I'm worry about compatibility of this case.

Reply:
I have no preference here.
Guys, do you know how other libraries work?
RFC compliance sounds a bold argument for me

Reply:
The most up to date RFC is https://tools.ietf.org/html/rfc2046, but I agree it specifies that there must be at least one body part.
I would argue however that this RFC does not allow representing a multipart body with *no* body part, so we can't follow the RFC in this case: adding an open-close boundary would be parsed as "single empty body part" by any server - which is of course the expected behaviour - so imho it is not an option here.

I definitely agree with @asvetlov that aiohttp should mimic existing librairies here:
 - urllib3 (and requests) serialise empty body with a lone closing boundary: https://github.com/urllib3/urllib3/blob/master/src/urllib3/filepost.py#L63
 - the `email.message` module in the stdlib does indeed follow the RFC: https://github.com/python/cpython/blob/master/Lib/email/generator.py#L291-L304 . But the consequence of this is that a "serialize-parse" loop with the stdlib implementation is not a no-op…